### PR TITLE
Lambdaauth

### PR DESCRIPTION
**Pull request by Tolledo**
*Issue #, if available:*

*Description of changes:*
On Sept 9th, API Gateway launched support for customers to secure Amazon API Gateway HTTP APIs using Lambda authorizers. Adding support for Lambda Authorizers to AWS Serverless Application Model.

*Description of how you validated changes:*
- updated tests

*Checklist:*

- [x] Write/update tests
- [x] `make pr` passes
- [ ] Update documentation
- [ ] Verify transformed template deploys and application functions as expected

*Examples?*

Please reach out in the comments, if you want to add an example. Examples will be 
added to `sam init` through https://github.com/awslabs/aws-sam-cli-app-templates/

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pytest.ini [hunk 1]
```diff
@@ -1,7 +1,7 @@
 [pytest]
 # Fail if coverage falls below 95%
 # NOTE: If debug breakpoints aren't working, comment out the code coverage line below
-addopts = --cov samtranslator --cov-report term-missing --cov-fail-under 95
+#addopts = --cov samtranslator --cov-report term-missing --cov-fail-under 95
```
Comment (wchengru) [inline pytest.ini:4]: Please revert the change if this line breaks your local test.

File: samtranslator/__init__.py [hunk 2]
```diff
@@ -1 +1 @@
-__version__ = "1.26.0"
+__version__ = "1.27.0"
```
Comment (wchengru) [inline samtranslator/__init__.py:1]: Please pull from develop branch and submit the pr again.

File: samtranslator/model/apigatewayv2.py [hunk 3]
```diff
@@ -57,39 +58,160 @@ class ApiGatewayV2ApiMapping(Resource):
...
+        function_arn=None,
+        function_invoke_role=None,
+        identity=None,
+        authorizer_payload_format_version=None,
+        enable_simple_responses=None,
     ):
         """
         Creates an authorizer for use in V2 Http Apis
         """
-        if authorization_scopes is not None and not isinstance(authorization_scopes, list):
```
Comment (wchengru) [inline samtranslator/model/apigatewayv2.py:65]: Why we delete this vaildation, and the validations below?

File: samtranslator/model/apigatewayv2.py [hunk 4]
```diff
@@ -57,39 +58,160 @@ class ApiGatewayV2ApiMapping(Resource):
...
+
+        if authorizer_type == "REQUEST":
+            if not self.function_arn:
+                raise InvalidResourceException(api_logical_id, name + " Lambda Authorizer must define 'FunctionArn'.")
+            if not self.authorizer_payload_format_version:
+                raise InvalidResourceException(
+                    api_logical_id, name + " Lambda Authorizer must define 'AuthorizerPayloadFormatVersion'."
+                )
+
+    def _get_type(self):
```
Comment (wchengru) [inline samtranslator/model/apigatewayv2.py:111]: Suggest to use a better function name, like _get_auth_type()

File: samtranslator/model/apigatewayv2.py [hunk 5]
```diff
@@ -57,39 +58,160 @@ class ApiGatewayV2ApiMapping(Resource):
...
-        openapi = {
-            "type": self.auth_type,
-            "x-amazon-apigateway-authorizer": {
-                "jwtConfiguration": self.jwt_configuration,
-                "identitySource": self.id_source,
-                "type": "jwt",
-            },
-        }
+        authorizer_type = self._get_type()
+        APIGATEWAY_AUTHORIZER_KEY = "x-amazon-apigateway-authorizer"
```
Comment (wchengru) [inline samtranslator/model/apigatewayv2.py:121]: Suggest to make the constant global, and put it at the top of file.

File: samtranslator/model/apigatewayv2.py [hunk 6]
```diff
@@ -57,39 +58,160 @@ class ApiGatewayV2ApiMapping(Resource):
...
+        if authorizer_type == "JWT":
+            openapi = {
+                "type": "oauth2",
+                "x-amazon-apigateway-authorizer": {
```
Comment (wchengru) [inline samtranslator/model/apigatewayv2.py:125]: minor: openapi[APIGATEWAY_AUTHORIZER_KEY] = {...} is more consistent with another branch.

File: samtranslator/model/apigatewayv2.py [hunk 7]
```diff
@@ -57,39 +58,160 @@ class ApiGatewayV2ApiMapping(Resource):
...
+            reauthorize_every = self._get_reauthorize_every()
+            if reauthorize_every is not None:
+                openapi[APIGATEWAY_AUTHORIZER_KEY]["authorizerResultTtlInSeconds"] = reauthorize_every
+
+            # Set identitySource if present
+            if self.identity:
+                openapi[APIGATEWAY_AUTHORIZER_KEY]["identitySource"] = self._get_identity_source()
+
+            # Set authorizerPayloadFormatVersion. It's a required parameter
+            openapi[APIGATEWAY_AUTHORIZER_KEY][
```
Comment (wchengru) [inline samtranslator/model/apigatewayv2.py:166]: minor: Can this line be reformatted?

File: samtranslator/model/iam.py [hunk 8]
```diff
@@ -23,7 +23,11 @@ def construct_assume_role_policy_for_service_principal(cls, service_principal):
         document = {
             "Version": "2012-10-17",
             "Statement": [
-                {"Action": ["sts:AssumeRole"], "Effect": "Allow", "Principal": {"Service": [service_principal]},}
+                {
```
Comment (wchengru) [inline samtranslator/model/iam.py:26]: Just to confirm: You didn't make any change on this file, right?

File: tests/translator/test_translator.py [hunk 9]
```diff
@@ -298,7 +298,6 @@ class TestTranslatorEndToEnd(TestCase):
                 "state_machine_with_api_resource_policy",
                 "state_machine_with_api_auth_default_scopes",
                 "state_machine_with_condition_and_events",
-                "state_machine_with_xray",
```
Comment (wchengru) [inline tests/translator/test_translator.py:301]: Why this line is deleted?

**Review by wchengru — CHANGES_REQUESTED** [review]

File: samtranslator/model/apigatewayv2.py (hunk 3 above)
Comment (Tolledo) [reply samtranslator/model/apigatewayv2.py:65]: I've just moved it further down, after we initialize all variables

File: tests/translator/test_translator.py (hunk 9 above)
Comment (Tolledo) [reply tests/translator/test_translator.py]: Seems like a merge from master. Reverted

File: samtranslator/model/iam.py (hunk 8 above)
Comment (Tolledo) [reply samtranslator/model/iam.py]: Yes, just formatting. Reverted back

File: samtranslator/__init__.py [hunk 10]
```diff
@@ -1 +1 @@
-__version__ = "1.27.0"
+__version__ = "1.26.0"
```
Comment (wchengru) [inline samtranslator/__init__.py:1]: Can you discard this change?

**Review by wchengru — CHANGES_REQUESTED** [review]

File: samtranslator/__init__.py (hunk 10 above)
Comment (Tolledo) [reply samtranslator/__init__.py]: Oops.

File: samtranslator/model/apigatewayv2.py [hunk 11]
```diff
@@ -57,39 +59,156 @@ class ApiGatewayV2ApiMapping(Resource):
...
+        self.function_invoke_role = function_invoke_role
+        self.identity = identity
+        self.authorizer_payload_format_version = authorizer_payload_format_version
+        self.enable_simple_responses = enable_simple_responses
+
+        # Validate parameters
+        if authorization_scopes is not None and not isinstance(authorization_scopes, list):
+            raise InvalidResourceException(api_logical_id, "AuthorizationScopes must be a list.")
+
+        # Validate necessary parameters exist
```
Comment (awood45) [inline samtranslator/model/apigatewayv2.py:92]: We also should validate that invalid parameters are not set (set to `None`) - so any parameters that are valid for only one auth type should be cause for rejection and an error if set for another auth type.

**Review by awood45 — CHANGES_REQUESTED** [review]
Will reach out with additional test changes.

File: samtranslator/model/apigatewayv2.py [hunk 12]
```diff
@@ -57,39 +59,156 @@ class ApiGatewayV2ApiMapping(Resource):
...
+
+        if authorizer_type == "REQUEST":
+            if not self.function_arn:
+                raise InvalidResourceException(api_logical_id, name + " Lambda Authorizer must define 'FunctionArn'.")
+            if not self.authorizer_payload_format_version:
+                raise InvalidResourceException(
+                    api_logical_id, name + " Lambda Authorizer must define 'AuthorizerPayloadFormatVersion'."
+                )
+
+    def _get_auth_type(self):
```
Comment (awood45) [inline samtranslator/model/apigatewayv2.py:112]: Now that there is more than one valid type of authorization, I'd prefer to avoid check/default - is any configuration without `jwt_configuration` a valid "REQUEST" authorizer? Or, are there other parameters that need to be present and we should add a third case of "This isn't valid at all, and we're going to fail with a helper message"?

**Review by awood45 — CHANGES_REQUESTED** [review]

File: samtranslator/model/apigatewayv2.py (hunk 12 above)
Comment (Tolledo) [reply samtranslator/model/apigatewayv2.py:112]: There is no assumption that any authorizer is valid by default. We have checks on lines 104-110 to determine if all required parameters for `REQUEST` authorizer are present.

If we add checks mentioned in your previous comment to validate that certain parameters only valid for one auth type - we should cover all scenarios.

File: samtranslator/model/apigatewayv2.py (hunk 11 above)
Comment (Tolledo) [reply samtranslator/model/apigatewayv2.py]: Added more validation + updated tests.

File: samtranslator/model/apigatewayv2.py (hunk 12 above)
Comment (awood45) [reply samtranslator/model/apigatewayv2.py:99]: Exactly right - we just want to make sure we check the different contingent conditions (which I see has changed) and fail with helpful error messages when appropriate.

**Review by awood45 — APPROVED** [review]
Looks good - though I see tests are failing, so let's get the tests passing before we can merge.

**Review by wchengru — APPROVED** [review]

**Comment by ivo** [comment]
Hi, not sure if this is the right place to ask. Although it's possible to setup lambda authorizers for httpAPI via the AWS console. But if I follow the SAM documentation I end up getting a internal server error unless I detach and re-attach the authorizer manually again.
Can you please share a working SAM template that was used to verify this feature? Thanks!
